### ISSUE TITLE
Perf issues on null-ls-backed actions

### ISSUE THREAD

**Issue by VVKot** [jose-elias-alvarez/nvim-lsp-ts-utils#35]
I just configured ESLint actions with null-ls and they seems to work like charm, thanks for that! I am noticing a lot of general slowness though (e.g. if I moving the cursor by spamming j/k it moves much slower than usual and there is similar lag on any other keypresses). When debugging #33 I noticed that we send lots of code actions request (I think when I move cursor to a specific line, which makes sense). I am also hitting a lot of "One shot channel can only send once" errors from plenary.

**Comment by jose-elias-alvarez** [comment]
That's really unusual – it should only send code action requests when you specifically request them, and it definitely shouldn't be doing anything when you move the cursor. I haven't seen anything like what you've mentioned in my own usage, either, but it sounds like something is spawning processes when it shouldn't. 

Could you post your `tsserver` config, including your null-ls config, so I can take a look and see if anything is going on there?

**Comment by VVKot** [comment]
Sure, I can get those. I think I know what's going on btw, that's nvim-lightbulb that does those requests for actions.

**Comment by VVKot** [comment]
Yeah, that's it, disabling the lightbulb fixes the issue. Maybe you have an idea on how I can do that better, e.g. is there a different event that wouldn't fire as often, or native vim way to debounce this call? Here's what I am current doing:

`vim.cmd [[autocmd CursorHold,CursorHoldI * lua require'nvim-lightbulb'.update_lightbulb()]]`

// understand that probably a better a way is to put some debouncing into nvim-lightbulb, just looking for a quick fix.

**Comment by jose-elias-alvarez** [comment]
By coincidence, I recently added a feature to null-ls to cache command output and added it to this plugin's integrations, which *should* theoretically handle cases like nvim-lightbulb's constant updates without hurting performance too badly. Are you using the latest version of null-ls? If not, could you try updating and see if that makes a difference?

Otherwise, it's a tricky situation, since as far as I know there's no built-in event that will behave better in this situation. If actual LSP behavior is OK, I don't think the nvim-lightbulb maintainers would want to change their plugin, either, so I might have to set up null-ls to handle this particular plugin.

**Comment by jose-elias-alvarez** [comment]
That's right, you should see the `spawning command` message only once, regardless of how many times you request code actions, until you modify the file, at which point the cache will be cleared until the command is spawned again (which is almost immediately with nvim-lightbulb).  

I tried setting up nvim-lightbulb on my end and it seems to be working as expected (i.e. it's only spawning once per change and using cached results otherwise) so I wonder what the issue is. Are you using any other LSP-related plugins that might be causing conflicts? It's possible that one of them is triggering a request that's clearing the cache.

I also just pushed a small change to null-ls to check the method more carefully before clearing the cache, so could you update and let me know if that makes a difference?

**Comment by jose-elias-alvarez** [comment]
Thanks for checking. I noted that the cache was storing the full path to the command, so I pushed another commit to strip the path and prevent any interference from, say, non-alphanumeric characters in the path name, but I don't know if that's the solution to this problem. 

I suspect it's some kind of incompatibility, so I put together this [minimal repo](https://github.com/jose-elias-alvarez/eslint-cache-minimal) to see if we can rule out null-ls itself. Could you try cloning it, running `nvim -u minimal.vim test.js`, and checking `:messages` to see if the command is spawning repeatedly?

**Comment by VVKot** [comment]
I am incredibly sorry. Seems that I messed up something in the nvim-lsp-ts-utils while trying to fix the problem I had earlier (with react types not being handled). I tried your repro on another machine and it worked just fine. So then I started looking what might be up, so that plugin's repo was dirty, and instinctively dropped changes&pulled.

After that things are working as expected (e.g. caching works) - no action needed! Sadly can't tell you what exactly I broke with regards to plugin to cause that behavior - but in any case you (likely) won't see people change plugin sources in the wild. 😄
